src/H5T.c

```diff
@@ -5056,6 +5065,60 @@ H5T_cmp(const H5T_t *dt1, const H5T_t *dt2, bool superset)
     FUNC_LEAVE_NOAPI(ret_value)
 } /* end H5T_cmp() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5T__bsearch_path_table
+ *
+ * Purpose:    Performs a binary search on the type conversion path table.
+ *             If `last_cmp` is non-NULL, the value of the last comparison
+ *             is returned through it. If `idx` is non-NULL, the idx into
+ *             the path table where the matching path was found is returned
+ *             through it. If no matching path is found, the value for
+ *             `idx` will be the index into the path table where a path
+ *             entry with source and destination datatypes matching src and
+ *             dst should be inserted. In this case, the caller should be
+ *             sure to increment the index value by 1 if the value of the
+ *             last comparison is > 0.
+ *
+ * Return:    Success:    Pointer to the path in the path table
+ *            Failure:    NULL if no matching path is found in the table
+ *
+ *-------------------------------------------------------------------------
+ */
+static void *
+H5T__bsearch_path_table(const H5T_t *src, const H5T_t *dst, int *last_cmp, int *idx)
```

Comment:
Moved the binary search logic into this helper function. I wanted to just implement this using `bsearch`, but its interface is a bit awkward to use for the things we're interested in knowing about like the last comparison value and the index into the table where we landed.